src/dataset.ts

```diff
@@ -220,33 +223,34 @@ export function checkStructures(o: any): [number, number] {
 
         if ('names' in structure && 'x' in structure && 'y' in structure && 'z' in structure) {
             for (const key of ['names', 'x', 'y', 'z']) {
-                if (!(key in structure && structure[key].length !== undefined)) {
+                if (!(key in structure)) {
                     throw Error(`missing '${name}' for structure ${i}`);
                 }
-                if (structure[key].length !== structure.size) {
-                    throw Error(`wrong size for '${name}' in structure ${i}, expected ${structure.size}, got ${structure[name].length}`);
+                const array = structure[key] as JsObject;
+
+                if (array.length !== structure.size) {
+                    throw Error(`wrong size for '${name}' in structure ${i}, expected ${structure.size}, got ${array.length}`);
                 }
             }
 
             if ('cell' in structure) {
-                if (structure.cell.length !== 9) {
+                if ((structure.cell as JsObject).length !== 9) {
                     throw Error(`'cell' must be an array of size 9 for structure ${i}`);
                 }
             }
         } else if ('data' in structure) {
             // nothing to check
         } else {
             throw Error(`structure ${i} must contains either 'names'/'x'/'y'/'z' or 'data'`);
-
         }
     }
 
     return [o.length, envCount];
 }
 
-function checkProperties(properties: any, structureCount: number, envCount: number) {
+function checkProperties(properties: JsObject, structureCount: number, envCount: number) {
```

Comment:
Could we instead have properties be a list of JsObjects, then the casting below becomes unnecessary?

Reply:
properties is an object, so we can not have a list here

```js
properties = {
    "KPCA[1]": {...},
    "KPCA[2]": {...},
     energy: {...} 
}
```

Reply:
Ah, so could it be a map of JSObjects?

Reply:
Could be a `Record<string, JsObject>` indeed. I'll change this